roachtest: sqlsmith/setup=seed/setting=no-mutations failed

[(roachtest).sqlsmith/setup=seed/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2034317&tab=buildLog) on [master@e85ac5f2d70542523a61f430f1fc354023b6c7e4](https://github.com/cockroachdb/cockroach/commits/e85ac5f2d70542523a61f430f1fc354023b6c7e4):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: unsupported comparison: bytes to jsonb
		stmt:
		SELECT
			(-1.9208989143371582):::FLOAT8 AS col_15025, tab_6146._jsonb AS col_15026
		FROM
			defaultdb.public.seed@seed__int8__float8__date_idx AS tab_6146
		WHERE
			(('1':::JSONB IS NOT DISTINCT FROM tab_6146._jsonb) AND true)
		LIMIT
			78:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2034317&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations)
Related:
- #49251 roachtest: sqlsmith/setup=seed/setting=no-mutations failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=seed/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2037483&tab=buildLog) on [master@4acd262796288cc84cb2b104bd77c1fc8cac73db](https://github.com/cockroachdb/cockroach/commits/4acd262796288cc84cb2b104bd77c1fc8cac73db):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: unsupported comparison: bytes to jsonb
		stmt:
		SELECT
			count(*) AS col_16303
		FROM
			defaultdb.public.seed@[0] AS tab_6636
		WHERE
			(e'[{"$=?^wDW?pd": [{"foo": []}], "T] i\'": 0.09139586892357388, "~xx>\'3t<3P": {}}, {"==I\\"zh,": {}, "Ay,o^": null, "b": [[true], true]}, []]':::JSONB IS NOT DISTINCT FROM tab_6636._jsonb)
		GROUP BY
			tab_6636._int8, tab_6636._float8
		LIMIT
			60:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2037483&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations)
Related:
- #49251 roachtest: sqlsmith/setup=seed/setting=no-mutations failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=seed/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2041122&tab=buildLog) on [master@d3791a81c0716478de08d44459d3fcf5b4f3ea1e](https://github.com/cockroachdb/cockroach/commits/d3791a81c0716478de08d44459d3fcf5b4f3ea1e):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: column 18 not in input
		stmt:
		WITH
			with_30 (col_157)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(((-9223372036854775807):::INT8::INT8 // 221.1264789855284273:::DECIMAL::DECIMAL)::DECIMAL),
								((-40995347891619.63925):::DECIMAL),
								(80016567882537.21702:::DECIMAL),
								((-47094.21070927867655):::DECIMAL)
						)
							AS tab_72 (col_157)
				)
		SELECT
			tab_73._enum AS col_158,
			tab_73._inet AS col_159,
			'\xf3':::BYTES AS col_160,
			NULL AS col_161,
			'00:00:00':::INTERVAL AS col_162,
			tab_73._date AS col_163,
			tab_73._uuid AS col_164,
			tab_73._timestamptz AS col_165,
			(-309479799747869998):::INT8 AS col_166,
			tab_73._jsonb AS col_167
		FROM
			defaultdb.public.seed@seed__int8__float8__date_idx AS tab_73
		ORDER BY
			tab_73._enum DESC, tab_73._date DESC, tab_73._uuid DESC
		LIMIT
			75:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2041122&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations)
Related:
- #49251 roachtest: sqlsmith/setup=seed/setting=no-mutations failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=seed/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2044000&tab=buildLog) on [master@8f768ad14cfb3f514db6d40465b2dd60ee1f2890](https://github.com/cockroachdb/cockroach/commits/8f768ad14cfb3f514db6d40465b2dd60ee1f2890):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:757: error: pq: internal error: runtime error: index out of range [0] with length 0
		stmt:
		WITH
			with_585 (col_3268)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								('0101000020E6100000000000000000F03F000000000000F03F':::GEOGRAPHY),
								('0101000020E6100000000000000000F03F000000000000F03F':::GEOGRAPHY),
								(NULL),
								('0101000020E6100000000000000000F03F000000000000F03F':::GEOGRAPHY),
								('0101000020E6100000000000000000F03F000000000000F03F':::GEOGRAPHY),
								(
									st_project('0101000020E6100000000000000000F87F000000000000F87F':::GEOGRAPHY::GEOGRAPHY, 1.0:::FLOAT8::FLOAT8, 0.260111032661812:::FLOAT8::FLOAT8)::GEOGRAPHY
								)
						)
							AS tab_1398 (col_3268)
				)
		SELECT
			'0101000020E6100000000000000000F03F000000000000F03F':::GEOGRAPHY AS col_3269
		FROM
			with_585 AS cte_ref_180
		WHERE
			false
		ORDER BY
			cte_ref_180.col_3268 DESC, cte_ref_180.col_3268, cte_ref_180.col_3268 ASC;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2044000&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations)
Related:
- #49251 roachtest: sqlsmith/setup=seed/setting=no-mutations failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=seed/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2045486&tab=buildLog) on [master@c627e3490d30e8ba88f6c7136717a392a054da4e](https://github.com/cockroachdb/cockroach/commits/c627e3490d30e8ba88f6c7136717a392a054da4e):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:757: error: pq: internal error: unsupported comparison: bytes to jsonb
		stmt:
		SELECT
			tab_12868._timestamp AS col_31329
		FROM
			defaultdb.public.seed@seed__int8__float8__date_idx AS tab_12867
			INNER JOIN defaultdb.public.seed@[0] AS tab_12868 ON true
		WHERE
			('null':::JSONB IS NOT DISTINCT FROM tab_12867._jsonb)
		ORDER BY
			tab_12867._bytes DESC, tab_12868._uuid ASC;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2045486&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations)
Related:
- #49251 roachtest: sqlsmith/setup=seed/setting=no-mutations failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=seed/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2046708&tab=buildLog) on [master@17c8048e80935f8a01477416980d18bf39cba1bb](https://github.com/cockroachdb/cockroach/commits/17c8048e80935f8a01477416980d18bf39cba1bb):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:757: error: pq: internal error: unsupported comparison: bytes to jsonb
		stmt:
		WITH
			with_6793 (col_39735) AS (SELECT * FROM (VALUES (166478304:::OID), (NULL)) AS tab_16273 (col_39735))
		SELECT
			tab_16274._enum AS col_39736, tab_16274._float4 AS col_39737
		FROM
			defaultdb.public.seed@seed__int8__float8__date_idx AS tab_16274
		WHERE
			(tab_16274._jsonb = e'{"=%(MH": 0.11662292965374893, "R+O": [{"baz": [], "foo": {"a": []}}, false], "baz": {}, "}0!!%SHF\'E&J": {}}':::JSONB)
		GROUP BY
			tab_16274._enum, tab_16274._float4
		HAVING
			min(tab_16274._bool::BOOL)::BOOL
		ORDER BY
			tab_16274._float4 ASC
		LIMIT
			81:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2046708&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations)
Related:
- #49251 roachtest: sqlsmith/setup=seed/setting=no-mutations failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/testdata/logic_test/inv_stats
+++ b/pkg/sql/logictest/testdata/logic_test/inv_stats
@@ -61,3 +61,10 @@ upper_bound             range_rows  distinct_range_rows  equal_rows
 '\xfd0fffffff00000000'  0           0                    1
 '\xfd1000000100000000'  0           0                    1
 '\xfd1aaaaaab00000000'  0           0                    1
+
+# Regression for #50596.
+statement ok
+CREATE TABLE t50596 (_jsonb JSONB, INVERTED INDEX (_jsonb));
+INSERT INTO t50596 VALUES ('{}');
+CREATE STATISTICS foo FROM t50596;
+SELECT NULL FROM t50596 WHERE '1' != _jsonb;

--- a/pkg/sql/opt/memo/statistics_builder.go
+++ b/pkg/sql/opt/memo/statistics_builder.go
@@ -18,6 +18,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/constraint"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/props"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util/json"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -700,7 +701,7 @@ func (sb *statisticsBuilder) colStatScan(colSet opt.ColSet, scan *ScanExpr) *pro
 	inputColStat := sb.colStatTable(scan.Table, colSet)
 	colStat := sb.copyColStat(colSet, s, inputColStat)
 
-	if sb.shouldUseHistogram(relProps) {
+	if sb.shouldUseHistogram(relProps, colSet) {
 		colStat.Histogram = inputColStat.Histogram
 	}
 
@@ -2484,11 +2485,21 @@ func (sb *statisticsBuilder) finalizeFromRowCountAndDistinctCounts(
 	}
 }
 
-func (sb *statisticsBuilder) shouldUseHistogram(relProps *props.Relational) bool {
+func (sb *statisticsBuilder) shouldUseHistogram(relProps *props.Relational, cols opt.ColSet) bool {
 	// If we know that the cardinality is below a certain threshold (e.g., due to
 	// a constraint on a key column), don't bother adding the overhead of
 	// creating a histogram.
-	return relProps.Cardinality.Max >= minCardinalityForHistogram
+	if relProps.Cardinality.Max < minCardinalityForHistogram {
+		return false
+	}
+	hasInv := false
+	cols.ForEach(func(col opt.ColumnID) {
+		colTyp := sb.md.ColumnMeta(col).Type
+		if sqlbase.ColumnTypeIsInvertedIndexable(colTyp) {
+			hasInv = true
+		}
+	})
+	return !hasInv
 }
 
 // rowsProcessed calculates and returns the number of rows processed by the
@@ -2863,7 +2874,7 @@ func (sb *statisticsBuilder) applyIndexConstraint(
 		sb.updateDistinctCountFromUnappliedConjuncts(col, e, s, numConjuncts, lowerBound)
 	}
 
-	if !sb.shouldUseHistogram(relProps) {
+	if !sb.shouldUseHistogram(relProps, constrainedCols) {
 		return constrainedCols, histCols
 	}
 
@@ -2926,12 +2937,12 @@ func (sb *statisticsBuilder) applyConstraintSet(
 			continue
 		}
 
-		if !sb.shouldUseHistogram(relProps) {
+		cols := opt.MakeColSet(col)
+		if !sb.shouldUseHistogram(relProps, cols) {
 			continue
 		}
 
 		// Calculate histogram.
-		cols := opt.MakeColSet(col)
 		inputStat, _ := sb.colStatFromInput(cols, e)
 		inputHist := inputStat.Histogram
 		if inputHist != nil {

